### PR TITLE
feat(engine): 8 new CLI commands — rand, note set, port list, clear, help (#99)

### LINKED ISSUE #99: [FEATURE] - New Commands for CLI
### Summary

| command | action |
|----------------|-----------|
| rand all | randomizes the notes and velocities in the sequencer |
| rand velo | randomizes only the velocities in the sequence |
| rand notes | randomizes only the notes in the sequence |
| note set {step} {note} | sets the note at sequence step to the specified step (velocity defaulted at 127 |
| note set {step} {note} {velo} | sets the note at sequence step to the specified step |
| port list | lists the available midi ports to connect too |
| clear | clears CLI |
| ok | clears CLI |
| help | lists available commands  and descriptions |

### Motivation

allowing a way for users to quickly set params using the cli. 

### Acceptance criteria

- [ ] refine

### Related items

_No response_

### PR DESCRIPTION
**Pull request by whinchman**
## Summary

Implements issue #99 — New Commands for CLI. Wires 8 new CLI commands into the engine across 5 files with full unit test coverage (167 tests passing).

- **`rand all` / `rand velo` / `rand notes`** — randomise all step notes+velocities, velocities only, or notes only (in-key) using the existing `rng_seed` via new `InputCommand::RandAll` and `InputCommand::RandVelocities` variants; `rand notes` reuses `GenerateRandomSequence`
- **`note set <step> <note_name> [velo]`** — sets note at a specific step accepting note names (C4, F#3, Bb2, A-1); velocity defaults to 127; dispatched via new `InputCommand::NoteSet { step, midi_note, velocity }`
- **`port list`** — fires `MidiCtrlMsg::ListPorts` to the MIDI output thread, which enumerates ALSA ports and replies on the log channel using a `[ports]` sentinel decoded by `run_ui`
- **`clear` / `ok`** — flushes `UiState.cli_log` directly with no `InputCommand`; **`help`** — prints all commands from a static `HELP_ENTRIES` constant

## Key Architectural Decisions

- `parse_note_name(s: &str) -> Option<u8>` added to `music_theory.rs` as a pure, no-alloc function — the inverse of the existing `note_name()`. Full test suite included.
- `port list` uses a fire-and-forget message pattern (`MidiCtrlMsg::ListPorts` → ALSA enumeration → `[ports]\x1F`-delimited sentinel back on `midi_log_rx`) rather than a cached `Vec<String>`, keeping `handle_cli_submit` side-effect-free with respect to ALSA.
- New `InputCommand` variants follow the existing single-writer invariant: `handle_cli_submit` never mutates `SequencerState` directly.
- `clear`/`ok` and `help` are handled entirely in `handle_cli_submit` without any `InputCommand`, consistent with the existing `port`/`channel`/`seed` patterns.

## Test Coverage

- `music_theory.rs`: 11 unit tests for `parse_note_name` (C4, case-insensitivity, sharps, flats, octave -1, out-of-range, empty, invalid letter, missing octave)
- `state.rs`: 4 unit tests for `RandAll`, `RandVelocities`, `NoteSet` (valid + out-of-range step)
- `ui.rs`: 19 unit tests for all new CLI branches — happy path, error cases, boundary conditions
- All external boundaries (ALSA, time) are mocked/stubbed via existing `run_midi_out_with_open_fn` and `make_test_ui` test helpers

## Known Limitations / Follow-up

- `port list` results arrive on the next render frame after the MIDI thread responds; imperceptible in practice but noted in the plan.
- MIDI note numbers (`note set 3 60`) are not accepted — name-only as specified in #99; can be added as a forward-compatible change.

Closes #99

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: engine/src/ui.rs [hunk 1]
```diff
@@ -180,6 +217,99 @@ pub(crate) fn handle_cli_submit(
...
+    ts: u64,
+    rest: &str,
+) {
+    let mut parts = rest.split_whitespace();
+
+    let step = match parts.next().and_then(|s| s.parse::<usize>().ok()) {
+        Some(s) if s <= 15 => s,
+        Some(_) => {
+            push_log(&mut ui.cli_log, ts, LogTag::Err, "step must be 0–15".into());
+            return;
```
[the remark this answers is not in the document]
Comment (whinchman) [reply engine/src/ui.rs]: Addressed in #111 (squash 8804376) — `note set` now accepts strict 1–16 (UI-facing), mapped to 0–15 internally. Error message updated to "step must be 1–16". `HELP_ENTRIES` entry rewritten to `("note set <1-16> <note> [vel]", …)`. Boundary tests added for steps 1 and 16; out-of-range tests assert 0 and 17 both error.

File: engine/src/ui.rs [hunk 2]
```diff
@@ -180,6 +217,99 @@ pub(crate) fn handle_cli_submit(
...
+                return;
+            }
+            Err(_) => {
+                push_log(&mut ui.cli_log, ts, LogTag::Err, format!("note set: invalid velocity '{s}'"));
+                return;
+            }
+        },
+        None => 127,
+    };
+
```
[the remark this answers is not in the document]
Comment (whinchman) [reply engine/src/ui.rs:315]: Addressed in #111 (squash 8804376) — `handle_cli_note_set` now rejects extra trailing tokens with `LogTag::Err` "note set: unexpected trailing input". New test `note_set_rejects_trailing_tokens` verifies zero `InputCommand`s are emitted on the rejection path.

File: engine/src/ui.rs [hunk 3]
```diff
@@ -180,6 +217,99 @@ pub(crate) fn handle_cli_submit(
...
+        let payload = msg.strip_prefix("[ports]")?;
+        if payload.is_empty() {
+            Some(vec![(LogTag::Info, "no MIDI ports available".into())])
+        } else {
+            Some(payload.split('\x1f').map(|name| (LogTag::Info, name.to_string())).collect())
+        }
+    } else if msg.starts_with("[ports-err]") {
+        Some(vec![(LogTag::Err, msg.to_string())])
+    } else {
+        None
```
[the remark this answers is not in the document]
Comment (whinchman) [reply engine/src/ui.rs:373]: Addressed in #111 (squash 8804376) — doc updated to match implementation (which was the intended behavior). Only `[ports-err]`-prefixed `ok=false` messages are wrapped as a sentinel `LogTag::Err`; any other `(false, _)` returns `None` so the caller (`run_ui` MIDI log drain) handles it via the generic `LogTag::Err` fall-through path. The function body is unchanged.

File: engine/src/ui.rs [hunk 4]
```diff
@@ -437,10 +567,20 @@ pub fn run_ui(
         // ── Drain MIDI log messages ───────────────────────────────────────────
         // Route messages from the MIDI output thread into the CLI log panel
         // before rendering so this frame shows the latest MIDI status.
+        // Special sentinels for port listing:
+        //   (true,  "[ports]name1\x1fname2") → one LogTag::Info per port name
+        //   (true,  "[ports]")               → "no MIDI ports available" LogTag::Info
+        //   (false, "[ports-err] ...")        → LogTag::Err (falls through to default)
```
[the remark this answers is not in the document]
Comment (whinchman) [reply engine/src/ui.rs]: Addressed in #111 (squash 8804376) — comment now reads `→ LogTag::Err via parse_ports_sentinel`, reflecting the actual control flow.

File: engine/src/midi_out.rs [hunk 5]
```diff
@@ -241,6 +243,22 @@ pub fn run_midi_out(
...
                 // This message is informational only for the midi_out thread.
             }
+            Ok(MidiCtrlMsg::ListPorts) => {
+                match midir::MidiOutput::new("midi-man-mk3-list") {
+                    Ok(output) => {
+                        let ports = output.ports();
+                        let names: Vec<String> = ports
+                            .iter()
+                            .map(|p| output.port_name(p).unwrap_or_default())
+                            .collect();
```
[the remark this answers is not in the document]
Comment (whinchman) [reply engine/src/midi_out.rs:280]: Addressed in #111 (squash 8804376) — when `output.port_name(p)` errors, the entry now falls back to `format!("port #{idx}")` (iteration index, not the port handle). Empty entries are filtered defensively before joining. New tests `list_ports_payload_uses_indexed_fallback_when_port_name_fails` and `list_ports_payload_contains_no_blank_entries` cover the fix.

File: engine/src/ui.rs [hunk 6]
```diff
@@ -34,8 +34,25 @@ use std::time::Instant;
...
+/// All CLI commands with brief descriptions shown by `help`.
+pub(crate) const HELP_ENTRIES: &[(&str, &str)] = &[
+    ("port <name>",      "connect to MIDI output port by name"),
+    ("port list",        "list available MIDI output ports"),
+    ("channel <1-16>",   "set MIDI output channel"),
+    ("seed <hex>",       "set random seed (e.g. 0xDEAD)"),
+    ("rand all",         "randomise notes and velocities"),
+    ("rand velo",        "randomise velocities only"),
+    ("rand notes",       "randomise note sequence"),
+    ("note set <step> <note> [vel]", "set a step's note and velocity"),
```
[the remark this answers is not in the document]
Comment (whinchman) [reply engine/src/ui.rs]: Addressed in #111 (squash 8804376) — `ok` alias added to `HELP_ENTRIES` as a sibling of `clear`. New tests `help_entries_includes_ok_alias` (asserts the entry exists in the constant) and `cli_submit_help_output_includes_ok_alias_line` (drives the rendered `help` output) cover both layers.
